### PR TITLE
WIP - documentation: generate API documentation programmatically

### PR DESCRIPTION
**Pull request by bill-auger**
this was discussed in chat - i would like to the API documentation; but these binary formats are tedious to review - this is a WIP/proof-of-concept for the sake of discussion - ive done enough to determine that this is not difficult to accomplish - the only caveat is that some new tools will need to be installed on the build server (pandoc and texlive - there is a simpler way to accomplish this by keeping the source an as HTML file; but experiments did not produce good results - the latex source is much more robust

### CODE REVIEW

**Comment by astralaster** [comment]
I've reverted this to draft, because its WIP.

**Comment by bill-auger** [comment]
ok, have made a lot of progress on this - i ended up with many questions though - i will upload the changes soon; but i may as well start asking questions now - these are two of the most significant, about which i could not describe well enough

------
for function 5,36 "Read Pixel", api.pdf has this description:
> Read Pixel at P01,P23 from the sprite layer (0-15, 0 = transparency)

it does not mention how to specify the layer - presumably it is Param4? - but more importantly, none of the other functions mention how a sprite becomes assigned to a layer - is there a yet undocumented function for that?

----
for function 6,2 "update sprite", api.pdf has this description:
> Image is P5 (bits 0-5 are sprite number, bit 6 indicates 32 bit – 
> NOT the same as the image number in the graphics system

yet in the source code (the SPRUpdate() function), Param5 is denoted as 'imageSize' - the commented printf trace indicates the same - indeed bit-6 is used to set the SPRITE_INTERNAL struct p->xSize p->ySize; but bits-0-5 are apparently used in GFXFindImage() as an index to locate the memory 

in other words, only bit-6 describes a size - bits-0-5 (documented as the "sprite number") could easily be described as "the image number in the graphics system"; but Param0 (the "sprite ID") could also be described as "the image number in the graphics system" - it is rather vague; so i am trying to clarify the terminology - instead of "image number" or "sprite number" i am trying to be explicit, more like "the index of the sprite in graphics memory" or "the index of the image in the graphics file"; but im not really sure what these parameters really are without digging deeper

can someone clarify, what is the difference between the "sprite ID" (Param0), the "sprite number" (Param5), and the vaguely mentioned "image number in the graphics system" (is that referring to Param0?)

**Comment by astralaster** [comment]
I don't think you should care about the content. This Pull request is about changing the process right? We can fix the content afterwards. :)

**Comment by bill-auger** [comment]
i have both motivations - my primary motivation is to understand how this
software works and how to program for it - the usual first step toward that is
to read the documentation - when i did that, i found it to be sparse,
inconsistent, and difficult to comprehend - so my next motivation was to
improve the documentation

however, reviewing changes to those binary documents is a PITA for the
maintainers; so the reason i am motivated to make the documentation easier to
maintain; is to increase the likelihood that my content improvements would be
accepted, and in the process, the many questions i had about the system will
get answered and anything missing from the docs will be accounted for

**Comment by paulscottrobson** [comment]
> for function 5,36 "Read Pixel", api.pdf has this description:
> > Read Pixel at P01,P23 from the sprite layer (0-15, 0 = transparency)
> it does not mention how to specify which layer, nor how to collect the result - presumably both are Param4? - or am i reading that wrong? - none of the other functions mention how a sprite becomes assigned to a layer - is there only one sprite layer, and "(0-15, 0 = transparency)" is telling how to interpret the "read pixel" function result ?

It returns the parameter in offset 0 of the buffer (most things do) which I normally refer to as "P0". There is no sprite layer really ; the system uses palette manipulation to create what is in practice a pair of 4 bit bitplanes. The sprite graphics are in the top nibble, the background in the bottom nibble, and the background only shows if the sprite graphic layer is zero. It's this top nibble that is read by 5,36. 

> for function 6,2 "update sprite", api.pdf has this description:
> > Image is P5 (bits 0-5 are sprite number, bit 6 indicates 32 bit –
> > NOT the same as the image number in the graphics system
> yet in the source code (the SPRUpdate() function), Param5 is denoted as 'imageSize' - the commented printf trace indicates the same - indeed bit-6 is used to set the SPRITE_INTERNAL struct p->xSize p->ySize; but bits-0-5 are apparently used in GFXFindImage() as an index to locate the memory
> 
> in other words, only bit-6 describes a size - bits-0-5 (documented as the "sprite number") could easily be described as "the image number in the graphics system"; but Param0 (the "sprite ID") could also be described as "the image number in the graphics system" - it is rather vague; so i am trying to clarify the terminology - instead of "image number" or "sprite number" i am trying to be explicit, more like "the index of this sprite abstraction in working memory" or "the index of the raw image data into the graphics file"; but im not sure what these parameters really are, without digging deeper
> 
> can someone clarify, what is the difference between the "sprite ID" (Param0), the "sprite number" (Param5), and the vaguely mentioned "image number in the graphics system" (is that referring to Param0?)

imageSize means 'image and size'. It's something of a fudge caused by the use of $80 to mean no change in sprite update, which meant I couldn't use $80 image number. In practice this value is the same as the sprite image number ($80-$BF for 16x16, $C0-$FF 32x32) just without bit 7 set. GFXFindImage takes the group (tile,sprite16,sprite32) and index, which is why it is anded with $3F

**Comment by astralaster** [comment]
@paulscottrobson If I understand it correctly this PR is about generating the PDF and ODT files on release from latex source. So it would be the same process to change documentation as to change code (which is optimal).

**Comment by paulscottrobson** [comment]
I don't mind either way. I've used Latex and Markdown ; swings and roundabouts. I agree this is the right solution ; I just wasn't sure if it was ready to be merged or not, or is it in development ?

**Comment by astralaster** [comment]
I think @bill-auger needs to push his latest changes. Current state is two weeks old.

**Review by paulscottrobson — APPROVED** [review]

**Comment by paulscottrobson** [comment]
This has been changed slightly by #230 as the build has changed slightly since it was pushed originally. I think I have kept the same idea.

**Comment by bill-auger** [comment]
i rebased and reworked yesterday (accounting for the new release/Makefile for example) - preparing to push now - it will probably conflict with that latest change; but i still can not compile this software, so i will try to work it out against the CI build

i considered other options such as HTML and markdown - both are brittle due to lack of standardization - i tried HTML first anyways; but could not format the tables properly with commonly available tools - markdown is just too limited for decent documentation - there is a reason why tex is used so widely by professionals - it is quirky, but reliable

> A major problem for me is I know how most of it works :) So it's quite difficult to see it from the POV of someone who doesn't and has to understand it.

i agree fully - i often tell people that the best people to write documentation are not the developers, but users, novices even (with input from the devs) - users are more likely to identify which parts are confusing or lacking
